In [ ]:
import pandas as pd
df = pd.read_csv('messy_sales.csv')
df.shape

(300, 6)

In [ ]:
df.head(10)

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [ ]:
df.dtypes

,0
order_id,int64
date,object
product,object
price,float64
qty,int64
zip,int64


Three problems I see with the data:
1. Date column has 2 different formats (yyyy-mm-dd and mm/dd/yyyy)
2. Some zip codes are wrong
3. Date column is not under the date data type

checking for missing data

In [ ]:
df.isna().sum()

,0
order_id,0
date,0
product,0
price,12
qty,0
zip,0


In [104]:
fill_value = df['price'].median()
df['price'] = df['price'].fillna(fill_value)
print(fill_value)

37.53


verifying missing data is removed

In [ ]:
df.isna().sum()

,0
order_id,0
date,0
product,0
price,0
qty,0
zip,0


checking for duplicates

In [ ]:
df.shape

(300, 6)

In [ ]:
print(df.duplicated().sum())

8


verifying duplicates were dropped

In [ ]:
df = df.drop_duplicates()

In [ ]:
df.shape

(292, 6)

changing data types

In [ ]:
df['zip'] = df['zip'].astype(str).str.zfill(5)

In [ ]:
df["date"] = pd.to_datetime(df["date"], format="mixed")

Verifying data change

In [ ]:
df.dtypes

,0
order_id,int64
date,datetime64[ns]
product,object
price,float64
qty,int64
zip,object


In [ ]:
df.head(10)

,order_id,date,product,price,qty,zip
0,1254,2026-04-06,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,2026-05-06,mug,5.50,1,10001
3,1066,2026-05-30,notebook,37.53,1,98101
4,1114,2026-04-06,webcam,45.59,4,90405
5,1280,2026-04-22,pen set,37.53,4,30303
6,1204,2026-04-18,charger,8.41,2,02134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [ ]:
df.shape

(292, 6)

looking at negative values in the data

In [ ]:
df[df['qty'] < 0]

,order_id,date,product,price,qty,zip
202,1140,2026-04-18,webcam,38.69,-5,98101
262,1233,2026-05-09,desk lamp,22.64,-4,10001
297,1025,2026-04-27,keyboard,47.30,-2,02116


flagging the negative data as returns

In [ ]:
returns = df[df['qty'] < 0]
print(returns)

     order_id       date    product  price  qty    zip
202      1140 2026-04-18     webcam  38.69   -5  98101
262      1233 2026-05-09  desk lamp  22.64   -4  10001
297      1025 2026-04-27   keyboard  47.30   -2  02116


saving the cleaned data set as a csv to reuse

In [ ]:
df.to_csv('sales_clean.csv', index = False)

#Cleaning Log

1. checking for missing data

    There were 12 missing rows from the price column. Price is a needed value for sales data, so I replaced the missing data within the column with the median value of the prices. I used the median because using the mean may overinflate price
    
    The 12 prices were filled with 37.53

2.  check for duplicates
    
    Originally there were 300 rows loaded into the data, after checking for duplicates there were 8 duplicate rows. Duplicates overinflate the data and cause problems with calculations and analysis, so I removed them.
    
    300 rows loaded; 8 duplicates removed; there are now 292 rows in the data

3. changing data types

    The date and zip columns were under the worng data types, text for date and integer for zip. Also, there were different date formats within the that column. Zip codes cannot be listed as a number in the case they start with a 0 AND they must be 5 charcters long. Dates need to be standardized and formatted correctly for prosperity and to make analysis easier.
    
    Changed zip column from integer to string; made zip codes 5 characters long; changed date column from text to date; standardized dates

4. dealing with negative values

    There are some negative values in the quantity column, but we don't know what they represent. Since the data tells us the items sold from a specific store, I interpreted the negative data as item returns.

    negatives kept with the data; stored under a variable named returns

5. Cleaning data is important because any misstep such as keeping duplicates or removing the wrong columns could cause issues wsuch as having too much data, too little, or having answers that don't make sense.